File: webapp/src/main/scala/org/linnando/mixemulator/webapp/RegistersComponent.scala
package org.linnando.mixemulator.webapp

import angulate2.std._
import org.linnando.mixemulator.vm.Comparison
import org.linnando.mixemulator.vm.io.data.IOWord

@Component(
  selector = "registers",
  templateUrl = "webapp/src/main/resources/registers.component.html",
  styleUrls = @@@("webapp/src/main/resources/registers.component.css")
)
class RegistersComponent(virtualMachineService: VirtualMachineService) {
  private def machineState = virtualMachineService.machineState

  def registerSign(register: String): String =
    registerContent(register).map(w => if (w.negative) "-" else "+").getOrElse("+")

  def registerByte(register: String, pos: Int): Byte =
    registerContent(register).map(_.bytes(pos)).getOrElse(0)

  def registerContent(register: String): Option[IOWord] = register match {
    case "A" => machineState.map(_.getA)
    case "X" => machineState.map(_.getX)
    case "I1" => machineState.map(_.getI1)
    case "I2" => machineState.map(_.getI2)
    case "I3" => machineState.map(_.getI3)
    case "I4" => machineState.map(_.getI4)
    case "I5" => machineState.map(_.getI5)
    case "I6" => machineState.map(_.getI6)
    case "J" => machineState.map(_.getJ)
  }

  def overflow: Boolean = machineState.exists(_.getOV)

  def comparison: Int = machineState.map(_.getCMP match {
    case Comparison.LESS => -1
    case Comparison.EQUAL => 0
    case Comparison.GREATER => 1
  }).getOrElse(0)

  def timeElapsed: Int = machineState.map(_.getTimeCounter).getOrElse(0)

  def isHalted: Boolean = machineState.exists(_.isHalted)
}

File: webapp/src/app/registers/registers.component.ts
import { Component } from '@angular/core';
import { IOWord, IOWord$ } from '@mixemulator/lib';
import { VirtualMachineService } from '../virtual-machine.service';

@Component({
  selector: 'mix-registers',
  templateUrl: './registers.component.html',
  styleUrls: ['./registers.component.scss']
})
export class RegistersComponent {
  constructor(private virtualMachineService: VirtualMachineService) { }

  registerSign(register: string): string {
    return this.registerContent(register).negative ? '-' : '+';
  }

  registerByte(register: string, pos: number): number {
    return this.registerContent(register).bytes[pos];
  }

  registerContent(register: string): IOWord {
    switch (register) {
      case 'A':
        return this.machineState?.getA ?? IOWord$.empty();
      case 'X':
        return this.machineState?.getX ?? IOWord$.empty();
      case 'I1':
        return this.machineState?.getI1 ?? IOWord$.empty();
      case 'I2':
        return this.machineState?.getI2 ?? IOWord$.empty();
      case 'I3':
        return this.machineState?.getI3 ?? IOWord$.empty();
      case 'I4':
        return this.machineState?.getI4 ?? IOWord$.empty();
      case 'I5':
        return this.machineState?.getI5 ?? IOWord$.empty();
      case 'I6':
        return this.machineState?.getI6 ?? IOWord$.empty();
      case 'J':
        return this.machineState?.getJ ?? IOWord$.empty();
      default:
        return IOWord$.empty();
    }
  }

  private get machineState() {
    return this.virtualMachineService.machineState;
  }

  get overflow(): boolean {
    return this.machineState?.getOV ?? false;
  }

  get comparison(): number {
    return this.machineState?.getCMP?.signum ?? 0;
  }

  get timeElapsed(): number {
    return this.machineState?.getTimeCounter ?? 0;
  }

  get isHalted(): boolean {
    return this.machineState?.isHalted ?? false;
  }
}
